integration_tests/testsuite/deploy_app.py

```diff
@@ -76,17 +80,67 @@ def _record_latency_to_bigquery(deploy_latency, language, is_xrt):
 def deploy_app_and_record_latency(appdir, language, is_xrt):
     start_time = time.time()
 
-    version = deploy_app(None, None, appdir, None)
+    version, url = deploy_app(None, None, appdir, None, 'GAE')
 
     # Latency is in seconds round up to 2 decimals
     deploy_latency = round(time.time() - start_time, 2)
 
     # Store the deploy latency data to bigquery
     _record_latency_to_bigquery(deploy_latency, language, is_xrt)
-    return version
+    return version, url
+
+
+def deploy_app_gae(yaml):
+    logging.debug('Starting deploy to GAE')
+
+    deployed_version = test_util.generate_version()
+
+    # TODO: once sdk driver is published, use it here
+    deploy_command = ['gcloud', 'app', 'deploy', '--no-promote',
+                      '--version', deployed_version, '-q']
+    if yaml:
+        logging.info(yaml)
+        deploy_command.append(yaml)
+
+    test_util.execute_command(deploy_command, True)
+    return (deployed_version,
+            test_util.retrieve_url_for_version(deployed_version))
+
+
+def deploy_app_gke(yaml):
+    logging.debug('Starting deploy to GKE')
+
+    image_name = test_util.generate_gke_image_name()
+    service_name = test_util.generate_gke_service_name()
+    namespace = test_util.generate_namespace()
+
+    build_command = ['docker', 'build', '-t', image_name, '.']
+    test_util.execute_command(build_command, True)
+
+    push_command = ['gcloud', 'docker', '--', 'push', image_name]
+    test_util.execute_command(push_command, True)
+
+    # This command updates the kubeconfig file with credentials for the given
+    # GKE cluster. Essentially, points kubectl to our preconfigured cluster.
+    cred_command = ['gcloud', 'container', 'clusters',
+                    'get-credentials', constants.CLUSTER_NAME]
+    test_util.execute_command(cred_command, True)
 
+    namespace_command = ['kubectl', 'create', 'namespace', namespace]
+    test_util.execute_command(namespace_command, True)
 
-def deploy_app(base_image, builder_image, appdir, yaml):
+    deploy_command = ['kubectl', 'apply', '-f', '-',
+                      '--namespace', namespace]
+    test_util.execute_command(deploy_command, True,
+                              template.GKE_TEMPLATE.format(
+                                  service_name=service_name,
+                                  test_image=image_name))
+
+    return namespace, test_util.get_external_ip_for_cluster(service_name,
+                                                            namespace)
+
+
+def deploy_app(base_image, builder_image, appdir, yaml, environment):
```

Comment:
If most of these are none in the case of GKE, could we make them default to None? Or make this take an environment settings object?

Reply:
ah yeah good call, I like defaulting them to None since there are a few different cases that are checked and handled by the main deploy method